daemon/disk_usage.go

```diff
@@ -5,46 +5,75 @@ import (
 	"fmt"
 	"sync/atomic"
 
+	"github.com/docker/docker/api/server/router/system"
 	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/errdefs"
 )
 
 // SystemDiskUsage returns information about the daemon data disk usage
-func (daemon *Daemon) SystemDiskUsage(ctx context.Context) (*types.DiskUsage, error) {
+func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsageOptions) (*types.DiskUsage, error) {
 	if !atomic.CompareAndSwapInt32(&daemon.diskUsageRunning, 0, 1) {
 		return nil, fmt.Errorf("a disk usage operation is already running")
 	}
 	defer atomic.StoreInt32(&daemon.diskUsageRunning, 0)
 
-	// Retrieve container list
-	allContainers, err := daemon.Containers(&types.ContainerListOptions{
-		Size: true,
-		All:  true,
-	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to retrieve container list: %v", err)
+	var (
+		containers []*types.Container
+		images     []*types.ImageSummary
+		layersSize int64
+		volumes    []*types.Volume
+		err        error
+	)
+	if len(opts.ObjectTypes) == 0 {
+		opts.ObjectTypes = map[types.DiskUsageObject]struct{}{
+			types.ContainerObject: {},
+			types.ImageObject:     {},
+			types.VolumeObject:    {},
+		}
 	}
+	for typ := range opts.ObjectTypes {
```

Comment:
This loop adds some unnecessary indentation.
Instead can we do

```go
if opts.ObjectTypes[types.ContainerObject] {
    containers, err = ...
}
```

Reply:
Won't that make it a very longe `if / else if / else if / else if / else` ? Wondering if that makes it more readable.

Reply:
Why? If we need containers then get them, otherwise there shouldn't be anything to do.

Reply:
Ah, yes you're right. Didn't think straight; Was thinking "we don't return, so we must handle the rest in "else"

Reply:
If we don't iterate over map and we use it like a struct with booleans, we might as well just use a struct with booleans instead, which is simpler and less error-prone.
E.g. see original implementation: https://github.com/moby/moby/blob/310910eba3f1417fcddc091f98941afcefa23df4/daemon/disk_usage.go#L22-L57